3792 providing capability to alter source resource content before mdm matching (#3802)
hapifhir/hapi-fhir: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/docs/server_jpa_mdm/mdm_customizations.md

# MDM Customizations

> This section describes customization supported by MDM.

Reviewer: Fantastic to see our documentation improved like this!!!  Excellent initiative!

Author: All the praise and credit should go to GG.  He told me where to put it and what he wanted.

# Interceptors

MDM allows customization through interceptors.  Please refer to the [Interceptors](/hapi-fhir/docs/interceptors/interceptors.html) section of the documentation for further details and implementation guidelines.   

> ## MDM Preprocessing Pointcut

Reviewer 2: Ideally indicate which pointcut it is : `MDM_BEFORE_RESOURCE_CHECKED` or something like that.

edit: Ah I see you mention it below, disregard this.

> MDM supports a pointcut invocation right before it starts matching an incoming source resource against defined rules.  A possible use of the pointcut would be to alter a resource property(ies) with the intention of influencing the MDM matching and linking process.  It should be noted that any modifications to the source resource are NOT persisted to the supporting storage module.  Modifications performed within the pointcut will remain valid during MDM processing only.

Reviewer: ```suggestion
MDM supports a pointcut invocation right before it starts matching an incoming source resource against defined rules.  A possible use of the pointcut would be to alter a resource content with the intention of influencing the MDM matching and linking process.  Any modifications to the source resource are NOT persisted to the database.  Modifications performed within the pointcut will remain valid during MDM processing only.
```

Reviewer: hapi-fhir does not have modules

Author: done

## Example: Ignoring Matches on Patient Name

In a scenario where a patient was given a placeholder name(John Doe), it would be desirable to ignore a 'name' matching rule but allow matching on other valid rules like matching SSN or a matching address.  

The following provides a full implementation of an interceptor that prevents matching on a patient name when it detects  a placeholder value.  

```java

> {{snippet:file:hapi-fhir/hapi-fhir-jpaserver-mdm/src/test/java/ca/uhn/fhir/jpa/mdm/interceptor/PatientNameModifierMdmPreProcessingInterceptor.java}}

Reviewer: snippets FTW!
